src/core/thread/network_data_service.hpp

```diff
@@ -324,42 +333,78 @@ class Manager : public InstanceLocator, private NonCopyable
     {
     public:
         OT_TOOL_PACKED_BEGIN
-        struct ServiceData
+        class AddrData
         {
         public:
-            explicit ServiceData(const Ip6::Address &aAddress, uint16_t aPort)
-                : mServiceNumber(kDnsSrpUnicastServiceNumber)
-                , mAddress(aAddress)
+            static constexpr uint8_t kMinLength = sizeof(Ip6::Address) + sizeof(uint16_t); // Address and port.
+
+            AddrData(const Ip6::Address &aAddress, uint16_t aPort, uint8_t aVersion)
+                : mAddress(aAddress)
                 , mPort(BigEndian::HostSwap16(aPort))
+                , mVersion(aVersion)
             {
-                OT_UNUSED_VARIABLE(mServiceNumber);
             }
 
+            uint8_t             GetLength(void) const { return (mVersion == 0) ? kMinLength : sizeof(AddrData); }
             const Ip6::Address &GetAddress(void) const { return mAddress; }
             uint16_t            GetPort(void) const { return BigEndian::HostSwap16(mPort); }
+            uint8_t             GetVersion(void) const { return mVersion; }
+
+            static Error ParseFrom(const uint8_t *aData, uint8_t aLength, DnsSrpUnicastInfo &aInfo);
 
         private:
-            uint8_t      mServiceNumber;
             Ip6::Address mAddress;
             uint16_t     mPort;
+            uint8_t      mVersion;
+        } OT_TOOL_PACKED_END;
+
+        OT_TOOL_PACKED_BEGIN
+        class ServiceData
+        {
+        public:
+            static constexpr uint8_t kMinLength = sizeof(uint8_t) + AddrData::kMinLength;
+
+            ServiceData(const Ip6::Address &aAddress, uint16_t aPort, uint8_t aVersion)
+                : mServiceNumber(kDnsSrpUnicastServiceNumber)
+                , mAddrData(aAddress, aPort, aVersion)
+            {
+                OT_UNUSED_VARIABLE(mServiceNumber);
+            }
+
+            uint8_t GetLength(void) const { return sizeof(uint8_t) + mAddrData.GetLength(); }
+
+            static Error ParseFrom(const ServiceTlv &aServiceTlv, DnsSrpUnicastInfo &aInfo)
+            {
+                // Skip over `mServiceNumber` field (`uint8_t`)`
+                return AddrData::ParseFrom(aServiceTlv.GetServiceData() + sizeof(uint8_t),
+                                           aServiceTlv.GetServiceDataLength() - sizeof(uint8_t), aInfo);
```

Comment:
Do we know here for sure that length >= 1 ? ( and so a check isn't needed)
There might be zero-length Service Data field generated (as attack or for some future reason..?)

Reply:
Yes. We know this. Generally, we must first determine the service type (ensuring we have the "service number" field, reading it, and checking it). Only then (knowing the specific service number type) can we try to parse it as `DnsSrpUnicast::ServiceData`.

The related code for this is in `GetNextDnsSrpUnicastInfo()`:

```cpp
serviceData.InitFrom(serviceNumber);
aIterator.mServiceTlv =
    Get<Leader>().FindNextThreadService(aIterator.mServiceTlv, serviceData, NetworkData::kServicePrefixMatch);
```

Here is a breakdown of the code:

  - `serviceData` is of type `Data` (a sequence of bytes with a specified length).
  - We initialize it from a single `uint8_t` value (`serviceNumber`), so it becomes a byte array with length one and the "service number" value.
  - `FindNextThreadService(aIterator.mServiceTlv, serviceData, NetworkData::kServicePrefixMatch)` finds the next `ServiceTlv` matching the given `serviceData` blob. We pass `NetworkData::kServicePrefixMatch` to use "prefix match" mode (accepting it if additional bytes follow the provided `serviceData`). This method also provides `kServiceExtactMatch` mode, which performs a full and exact match (same length, same bytes).

Reply:
ok, clear!